pkg/collector/container_hc_collector.go

```diff
@@ -66,67 +146,44 @@ func (c *Collector) updateBPFMetrics() {
 		if err != nil {
 			klog.V(5).Infof("failed to resolve container for cGroup ID %v: %v, set containerID=%s", ct.CGroupID, err, c.systemProcessName)
 		}
-		// TODO: improve the removal of deleted containers from ContainersMetrics. Currently we verify the maxInactiveContainers using the foundContainer map
-		foundContainer[containerID] = true
 
 		c.createContainersMetricsIfNotExist(containerID, ct.CGroupID, ct.PID, config.EnabledEBPFCgroupID)
-
+		comm := C.GoString((*C.char)(unsafe.Pointer(&ct.Command)))
 		// System process is the aggregation of all background process running outside kubernetes
 		// this means that the list of process might be very large, so we will not add this information to the cache
 		if containerID != c.systemProcessName {
 			c.ContainersMetrics[containerID].SetLatestProcess(ct.CGroupID, ct.PID, comm)
 		} else if config.EnableProcessMetrics {
 			c.createProcessMetricsIfNotExist(ct.PID, comm)
-			if err = c.ProcessMetrics[ct.PID].CPUTime.AddNewDelta(ct.ProcessRunTime); err != nil {
+			if err := c.ProcessMetrics[ct.PID].CPUTime.AddNewDelta(ct.ProcessRunTime); err != nil {
 				klog.V(5).Infoln(err)
 			}
 		}
+		isSystemProcess := containerID == c.systemProcessName && config.EnableProcessMetrics
 
-		// update ebpf metrics
-		// first update CPU time
-		if err = c.ContainersMetrics[containerID].CPUTime.AddNewDelta(ct.ProcessRunTime); err != nil {
-			klog.V(5).Infoln(err)
-		}
-		// update IRQ vector
-		for i := 0; i < config.MaxIRQ; i++ {
-			if err = c.ContainersMetrics[containerID].SoftIRQCount[i].AddNewDelta(uint64(ct.VecNR[i])); err != nil {
-				klog.V(5).Infoln(err)
-			}
+		// set CPUIDs for specific pid
+		var cpuIDs map[int32]int
+		var ok bool
+		if cpuIDs, ok = processCPUMap[ct.PID]; !ok {
+			cpuIDs = make(map[int32]int)
 		}
-
-		// update HW counters
-		for _, counterKey := range collector_metric.AvailableHWCounters {
-			var val uint64
-			switch counterKey {
-			case attacher.CPUCycleLabel:
-				val = ct.CPUCycles
-			case attacher.CPUInstructionLabel:
-				val = ct.CPUInstr
-			case attacher.CacheMissLabel:
-				val = ct.CacheMisses
-			default:
-				val = 0
-			}
-
-			if err = c.ContainersMetrics[containerID].CounterStats[counterKey].AddNewDelta(val); err != nil {
-				klog.V(5).Infoln(err)
-			}
-			// track system process metrics
-			if containerID == c.systemProcessName && config.EnableProcessMetrics {
-				foundProcess[ct.PID] = true
-
-				if err = c.ProcessMetrics[ct.PID].CounterStats[counterKey].AddNewDelta(val); err != nil {
-					klog.V(5).Infoln(err)
-				}
-				for i := 0; i < config.MaxIRQ; i++ {
-					if err = c.ProcessMetrics[ct.PID].SoftIRQCount[i].AddNewDelta(uint64(ct.VecNR[i])); err != nil {
-						klog.V(5).Infoln(err)
+		if isSystemProcess {
+			// always 1
+			c.ProcessMetrics[ct.PID].CPUIDs = cpuIDs
+		} else {
+			if _, found := foundContainer[containerID]; !found {
+				// newly set to 1
+				c.ContainersMetrics[containerID].CPUIDs = cpuIDs
```

Comment:
CPUID means CPU counters or ID ? the name seems means CPU ID but actually it's number/count that run the process ?

Reply:
Changed the name to `processCountOnCPU`. 
Firstly, I just thought about exposing the CPU IDs but now it more like pushing the ID to label and have the process count as a value instead. 
Thanks.